### PR TITLE
Feat/goal likelihood metric

### PR DESCRIPTION
**Pull request by maharajamihir**

### CODE REVIEW

**Comment by emergenz** [comment]
We should turn `search_budget` into a config parameter, and then change the new parameter for `proof_search` etc. to be a Boolean `is_eval=False`.

Also, since the current validation loss does not really work, I propose completely replacing the current validation loss with this new one, hence I think we should calculate the MCTS-based logprobs after each iteration (instead of before).

File: learning/bootstrap.py [hunk 1]
```diff
@@ -109,11 +109,27 @@ async def teacher_loop(cfg: DictConfig, mle_log: MLELogger):
...
+                            'mean_success_logprobs_final': mean_success_logprobs_final,
                            {'final_goals_proven': final_goals_proven})
                 break
 
+            # get logprobs of proving the final goals (with far more mcts steps) as a progress_metric for us
+            search_budget = min(cfg.agent.max_mcts_nodes*100, 10000)
+            while len(success_logprobs_final) == 0 and search_budget < 100000:
+                student_results_final = prove_conjectures(agent_dump, final_goals_formatted, theory, premises, search_budget)
+                success_logprobs_final = get_log_probs(student_results_final, i)
+                search_budget *= 2
```
Comment (emergenz) [inline learning/bootstrap.py:124]: I'm not sure about the exponential backoff logic here. We should either constrain the MCTS search to timeout after `search_budget` steps, or we let the MCTS search run indefinitely, i.e. until proving the theorem.

Currently we would be re-running MCTS from zero when the exponential backoff logic is triggered, thus wasting a lot of compute.

File: learning/bootstrap.py [hunk 2]
```diff
@@ -109,11 +109,27 @@ async def teacher_loop(cfg: DictConfig, mle_log: MLELogger):
...
+
+            if len(success_logprobs_final) > 0:
+                mean_success_logprobs_final = sum(success_logprobs_final)/len(success_logprobs_final)
+            else:
+                mean_success_logprobs_final = -100
```
Comment (emergenz) [inline learning/bootstrap.py:129]: I would put this entire block into `get_val_loss`

File: learning/bootstrap.py [hunk 3]
```diff
@@ -217,6 +233,7 @@ async def teacher_loop(cfg: DictConfig, mle_log: MLELogger):
                 mle_log.update({'num_iterations': i},
                            {'val_loss': val_loss,
                             'final_goals_proven': final_goals_proven,
+                            'mean_success_logprobs_final': mean_success_logprobs_final,
```
[the remark this answers is not in the document]
Comment (emergenz) [reply learning/bootstrap.py:236]: nit: can you do `mean_success_log_probs_final` to adhere to our naming scheme?

File: learning/bootstrap.py [hunk 4]
```diff
@@ -229,15 +227,34 @@ async def teacher_loop(cfg: DictConfig, mle_log: MLELogger):
...
+    # get logprobs of proving the final goals (with far more mcts steps) as a progress_metric for us
+    student_results_final = prove_conjectures(agent_dump, final_goals_formatted, theory, premises, is_eval=True)
+    success_logprobs_final = get_log_probs(student_results_final, i)
+
+    if len(success_logprobs_final) > 0:
+        mean_success_logprobs_final = sum(success_logprobs_final)/len(success_logprobs_final)
+    else:
+        mean_success_logprobs_final = -10
+
+    # flip signs so it is a loss
```
Comment (emergenz) [inline learning/bootstrap.py:245]:
```suggestion
```

File: learning/bootstrap.py [hunk 5]
```diff
@@ -213,15 +201,21 @@ async def teacher_loop(cfg: DictConfig, mle_log: MLELogger):
...
-            if i + 1 < cfg.agent.policy.total_iterations:
-                print(len(examples), 'accumulated training examples.')
-                val_loss = agent.train(examples=examples, final_goals=final_goals, solutions=final_solutions, ratio_proven=ratio_proven, mle_log=mle_log)
-                mle_log.update({'num_iterations': i},
-                           {'val_loss': val_loss,
-                            'final_goals_proven': final_goals_proven,
-                            'ratio_proven': ratio_proven,
-                            'mean_hard_sol_log_probs': mean_hard_sol_log_prob},
-                            extra_obj={'conjectured_final_goals': conjectured_final_goals})
+            print(len(examples), 'accumulated training examples.')
```
Comment (emergenz) [inline learning/bootstrap.py:204]:
```suggestion
            log.info(f"{len(examples)} accumulated training examples.")
```

**Review by emergenz — APPROVED** [review]
